mzLib/Proteomics/Fragmentation/MatchedFragmentIon.cs

```diff
@@ -1,26 +1,45 @@
 ﻿using Chemistry;
 using System.Text;
+using System.Text.RegularExpressions;
+using MassSpectrometry;
 
 namespace Proteomics.Fragmentation
 {
     public class MatchedFragmentIon
     {
-        public readonly Product NeutralTheoreticalProduct;
+        public readonly IProduct NeutralTheoreticalProduct;
         public readonly double Mz;
         public readonly double Intensity;
         public readonly int Charge;
 
         /// <summary>
         /// Constructs a new MatchedFragmentIon given information about a theoretical and an experimental fragment mass spectral peak
         /// </summary>
-        public MatchedFragmentIon(ref Product neutralTheoreticalProduct, double experMz, double experIntensity, int charge)
+        public MatchedFragmentIon(IProduct neutralTheoreticalProduct, double experMz, double experIntensity, int charge)
         {
             NeutralTheoreticalProduct = neutralTheoreticalProduct;
             Mz = experMz;
             Intensity = experIntensity;
             Charge = charge;
         }
 
+        /// <summary>
+        /// Constructs a new MatchedFragmentIon given information about a theoretical and an experimental fragment mass spectral peak
+        /// </summary>
+        public MatchedFragmentIon(ref IProduct neutralTheoreticalProduct, double experMz, double experIntensity, int charge)
```

Comment:
I believe that the ref keyword is only applicable for reference types if you need to replace it entirely. e.g., the matched fragment ion constructor was creating a new IProduct, assigning it to neutralTheoreticalProduct, and you wanted that new assignment to be visible outside of the local type.

tldr; lose the ref